### PR TITLE
Adding tests, Interfaces

### PR DESCRIPTION
**Pull request by kuhlenh**
Need to remove User.cs, Trainer.cs, etc.

### CODE REVIEW

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+                case Gender.Female:
+                    return 655.1 + (9.563 * Weight * 0.453592) + (1.850 * Height * 2.54) - (4.676 * Age);
+                default:
+                    return 0.0;
+            }
+        }
+
+        public double GetCaloriesBurned(IWorkout workout)
+        {
+            //var a = (Gender)5;
```
Comment (jmarolf) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: >//var a = (Gender)5; [](start = 12, length = 20)

delete this

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+        double Distance { get; }
+        double Pace { get; }
+    }
+
+    public interface IWorkoutType
+    {
+        WorkoutType Type { get; }
+    }
+
+    public interface IBikeWorkout : IDistanceWorkout, IWorkout
```
Comment (jmarolf) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: >IBikeWorkout [](start = 21, length = 12)

You shouldn't need an interface that is just inherits from other interfaces.

File: src/BikeSharing.DomainLogic/Trainer.cs [hunk 3]
```diff
@@ -5,7 +5,7 @@
 
 namespace Training
 {
-    public class Workout
+    public class Workout2
```
Comment (jmarolf) [inline src/BikeSharing.DomainLogic/Trainer.cs]: >Workout2 [](start = 17, length = 8)

please come up with a better name

File: src/BikeSharing.DomainLogic.Tests/TrainerTests.cs [hunk 4]
```diff
@@ -1,4 +1,5 @@
-﻿using Microsoft.VisualStudio.TestTools.UnitTesting;
+﻿using BikeSharing.DomainLogic;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
 using System;
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic.Tests/TrainerTests.cs]: Sort system usings first.

File: src/BikeSharing.DomainLogic.Tests/TrainerTests.cs [hunk 5]
```diff
@@ -8,138 +9,355 @@ namespace Trainer.Tests
...
+    //    {
+    //        CreateFemaleAthlete();
+    //        var best = _user.GetBestWorkoutThisWeek();
+    //        var workout = new BikeWorkout(3, new TimeSpan(2, 16, 34), 151,
+    //                                      DateTime.Now.AddDays(-4), 14.27,
+    //                                      _user,
+    //                                      "Biking to Red Hook!");
+    //        Assert.AreEqual((workout, 3), best);
+    //    }
+    //}
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic.Tests/TrainerTests.cs]: remove commented out code.

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 6]
```diff
@@ -0,0 +1,328 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using BikeSharing.DomainLogic;
+
+namespace Training
+{
+
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: move blank line.

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 7]
```diff
@@ -0,0 +1,328 @@
+﻿using System;
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs:1]: Copyright.

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+using BikeSharing.DomainLogic;
+
+namespace Training
+{
+
+    public interface IWorkout
+    {
+        DateTime Date { get; }
+        TimeSpan Duration { get; }
+        IHeartRate HeartRate { get; }
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: IHeartRate?

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+    public interface IDistanceWorkout : IWorkout
+    {
+        double Distance { get; }
+        double Pace { get; }
+    }
+
+    public interface IWorkoutType
+    {
+        WorkoutType Type { get; }
+    }
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: this interface is weird... why isn't "Type" just part of IWorkout?  Why do you need this interface that just exposes a single member?

File: src/BikeSharing.DomainLogic/BikeWorkout.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/BikeSharing.DomainLogic/BikeWorkout.cs]: agreed.  waht does IBikeWorkout *add* that's interesting?  also, IDistanceWorkout is an IWorkout, so you don' tneed to list both.

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+            this.duration = duration;
+            this.rate = rate;
+            if (notes == null)
+            {
+                throw new ArgumentNullException(nameof(notes));
+            }
+            this.notes = notes;
+        }
+
+        public DateTime Date => date;
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: are you goign to convert these to auto-props in the demo?

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+        public TimeSpan Duration => duration;
+        public IHeartRate HeartRate => rate;
+        public string Notes { get => notes; set => notes = value; }
+    }
+
+    public class DistanceWorkout : IDistanceWorkout
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: whynot derive from Workout to save a lot on the implementation?

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+            {
+                case Gender.Male:
+                    return (-55.0969 + (0.6309 * workout.HeartRate.GetHeartRate()) 
+                           + (0.1988 * Weight) + (0.2017 * Age) / 4.184) 
+                           * 60 * workout.Duration.TotalHours;
+                case Gender.Female:
+                    return (-20.4022 + (0.4472 * workout.HeartRate.GetHeartRate()) 
+                           - (0.1263 * Weight) + (0.074 * Age) / 4.184) 
+                           * 60 * workout.Duration.TotalHours;
+                default:
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: ?

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+                    return 0.0;
+            }
+        }
+
+        public IWorkout GetWeeksBestWorkout()
+        {
+            var week = Workouts.Where(w => w.Date > DateTime.Now.Date.AddDays(-7));
+            return week.Aggregate((w1, w2) => w1.HeartRate.GetHeartRate() > w2.HeartRate.GetHeartRate() ? w1 : w2);
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: week.Max(...)

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+                           - (0.1263 * Weight) + (0.074 * Age) / 4.184) 
+                           * 60 * workout.Duration.TotalHours;
+                default:
+                    return 0.0;
+            }
+        }
+
+        public IWorkout GetWeeksBestWorkout()
+        {
+            var week = Workouts.Where(w => w.Date > DateTime.Now.Date.AddDays(-7));
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: var lastWeekWorkouts

File: src/BikeSharing.DomainLogic/Trainer.cs (hunk 3 above)
Comment (kuhlenh) [reply src/BikeSharing.DomainLogic/Trainer.cs]: oh i need to delete this class...

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+
+        public double GetCaloriesBurned(Workout workout)
+        {
+            //var a = (Gender)5;
+            switch (Gender)
+            {
+                case Gender.Male:
+                    return (-55.0969 + (0.6309 * workout.AverageHeartRate) 
+                           + (0.1988 * Weight) + (0.2017 * Age) / 4.184) 
+                           * 60 * workout.Duration.TotalHours;
```
Comment (jmarolf) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: I would add a function that makes this more obvious.  There are a lot of magic numbers here

File: src/BikeSharing.DomainLogic.Tests/TrainerTests.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/BikeSharing.DomainLogic.Tests/TrainerTests.cs]: Sort system usings first.

File: src/BikeSharing.DomainLogic.Tests/TrainerTests.cs [hunk 16]
```diff
@@ -8,138 +9,234 @@ namespace Trainer.Tests
 {
 
     [TestClass]
-    public class WorkoutTests
+    public class TestAthlete
     {
-        User _user;
+        Athlete athlete;
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic.Tests/TrainerTests.cs]: keeping around state in test classes in not recommended.  Just move this to be a local vairable in your test methods.

File: src/BikeSharing.DomainLogic/BikeWorkout.cs (hunk 7 above)
Comment (CyrusNajmabadi) [reply src/BikeSharing.DomainLogic/BikeWorkout.cs:1]: Copyright

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 17]
```diff
@@ -0,0 +1,258 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using BikeSharing.DomainLogic;
+
+namespace Training
+{
+    public class Workout
+    {
+        public DateTime Date { get; set; }
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: should this really be mutable?

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+        public TimeSpan Duration { get; set; }
+        public double AverageHeartRate { get; set; }
+        public string Notes;
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: why are the rest properties, and this is a field?

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 19]
```diff
@@ -0,0 +1,258 @@
...
+            {
+                throw new ArgumentNullException();
+            }
+            Notes = notes;
+        }
+    }
+
+    public class DistanceWorkout : Workout
+    {
+        public double Distance { get; set; }
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: why is this mutable?

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 20]
```diff
@@ -0,0 +1,258 @@
...
+        public double Pace { get; }
+        public DistanceWorkout(double distance, DateTime datetime, TimeSpan duration, double rate, string notes) : base(datetime, duration, rate, notes)
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: wrap :base call to make code clearer.

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 21]
```diff
@@ -0,0 +1,258 @@
...
+        {
+            Distance = distance;
+            Pace = distance / duration.TotalHours;
+        }
+    }
+
+    public class BikeWorkout : DistanceWorkout
+    {
+        public WorkoutType Type { get; set; }
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: mutable?

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 22]
```diff
@@ -0,0 +1,258 @@
...
+
+        public BikeWorkout(WorkoutType type, double distance, DateTime datetime, TimeSpan duration, double rate, string notes) : base(distance, datetime, duration, rate, notes)
+        {
+            Type = type;
+        }
+    }
+
+    public class Athlete
+    {
+        private readonly string[] HASHTAGS = new string[]{
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: consider hiding in a region so this doesn't clutter hte demo.

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 23]
```diff
@@ -0,0 +1,258 @@
...
+"#fitnessquote",
+"#youcandoit",
+"#justbringit",
+"#dreambig",
+"#success",
+"#staypositive",
+"#noexcuses"
+    };
+
+        public string Username { get; set; }
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: mutable?

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 24]
```diff
@@ -0,0 +1,258 @@
...
+        public List<Workout> Workouts { get; set; }
+
+        public Athlete(string username, Gender gender, int age, double weight, double height)
+        {
+            Username = username;
+            Gender = gender;
+            Age = age;
+            Weight = weight;
+            Height = height;
+            BasalMetabolicRate = GetBasalMetabolicRate();
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: why does this need to be stored in a field?  Why not just have the property compute the result when asked?

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 25]
```diff
@@ -0,0 +1,258 @@
...
+        }
+
+        public Athlete(string username, Gender gender, int age, double weight, double height, List<Workout> workouts)
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: why are there two constructs that do nearly the exact same thing?  why not have one call the other?

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 26]
```diff
@@ -0,0 +1,258 @@
...
+
+        private double GetBasalMetabolicRate()
+        {
+            switch (Gender)
+            {
+                case Gender.Male:
+                    return 66.47 + (13.75 * Weight * 0.453592) + (5.003 * Height * 2.54) - (6.755 * Age);
+                case Gender.Female:
+                    return 655.1 + (9.563 * Weight * 0.453592) + (1.850 * Height * 2.54) - (4.676 * Age);
+                default:
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: how would this happen?

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 27]
```diff
@@ -0,0 +1,258 @@
...
+                    return 0.0;
+            }
+        }
+
+        public double GetCaloriesBurned(Workout workout)
+        {
+            //var a = (Gender)5;
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: remove.

File: src/BikeSharing.DomainLogic/BikeWorkout.cs (hunk 18 above)
Comment (kuhlenh) [reply src/BikeSharing.DomainLogic/BikeWorkout.cs]: demo magic

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 28]
```diff
@@ -0,0 +1,258 @@
...
+
+        public string TweetTodaysWorkout()
+        {
+            if (Workouts != null)
+            {
+                var todaysWorkout = Workouts.Where(w => w.Date.Date == DateTime.Now.Date).FirstOrDefault();
+                if (todaysWorkout != null)
+                {
+                    var bike = todaysWorkout as BikeWorkout;
+                    if (bike != null)
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs:198]: i assume you're going to use the quick-fix suggestion here.

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 29]
```diff
@@ -0,0 +1,258 @@
...
+            }
+            return null;
+        }
+
+        private string Tweetify(string msg)
+        {
+            if (msg.Length >= 140)
+                return msg.Substring(0, 137) + "...";
+            else
+                return msg + GetHashTags(140 - msg.Length);
```
Comment (CyrusNajmabadi) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: lots of repeated, hardcoded values.  extract into one location.

File: src/BikeSharing.DomainLogic/BikeWorkout.cs (hunk 28 above)
Comment (kuhlenh) [reply src/BikeSharing.DomainLogic/BikeWorkout.cs:198]: yes

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 30]
```diff
@@ -0,0 +1,258 @@
...
+                           - (0.1263 * Weight) + (0.074 * Age) / 4.184) 
+                           * 60 * workout.Duration.TotalHours;
+                default:
+                    return 0.0;
+            }
+        }
+
+        public (Workout workout, double calories) GetWeeksBestWorkout()
+        {
+            var lastWeekWorkouts = Workouts.Where(w => w.Date > DateTime.Now.Date.AddDays(-7));
```
Comment (jmarolf) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: >DateTime.Now.Date.AddDays(-7)); [](start = 64, length = 31)

DateTime.Now.Date - TimeSpan.FromDays(7);

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 31]
```diff
@@ -0,0 +1,258 @@
...
+        public Workout(DateTime date, TimeSpan duration, double averageHeartRate, string notes)
+        {
+            Date = date;
+            Duration = duration;
+            AverageHeartRate = averageHeartRate;
+            if (notes == null)
+            {
+                throw new ArgumentNullException();
+            }
+            Notes = notes;
```
Comment (jmarolf) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs:24]: Notes = notes ?? new ArgumentNullException(nameof(notes));

Comment (kuhlenh) [reply src/BikeSharing.DomainLogic/BikeWorkout.cs:24]: yes this will be demo magic!

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 32]
```diff
@@ -0,0 +1,258 @@
...
+        }
+    }
+
+    public class DistanceWorkout : Workout
+    {
+        public double Distance { get; set; }
+        public double Pace { get; }
```
Comment (jmarolf) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: >public double Pace { get; } [](start = 8, length = 27)

public double Pace => Distance / Duration.TotalMinutes;

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 33]
```diff
@@ -0,0 +1,258 @@
...
+        public DistanceWorkout(double distance, DateTime datetime, TimeSpan duration, double rate, string notes) : base(datetime, duration, rate, notes)
+        {
+            Distance = distance;
+            Pace = distance / duration.TotalHours;
```
Comment (jmarolf) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: >TotalHours [](start = 39, length = 10)

who measures there time on hours? I want to know if I did a 5 minutes mile, not if I ran 20 miles per hour

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 34]
```diff
@@ -0,0 +1,258 @@
...
+            Username = username;
+            Gender = gender;
+            Age = age;
+            Weight = weight;
+            Height = height;
+            BasalMetabolicRate = GetBasalMetabolicRate();
+            Workouts = workouts;
+        }
+
+        public void AddWorkout(params Workout[] w)
```
Comment (jmarolf) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: >w [](start = 48, length = 1)

rename to workouts

File: src/BikeSharing.DomainLogic/BikeWorkout.cs (hunk 30 above)
Comment (jmarolf) [reply src/BikeSharing.DomainLogic/BikeWorkout.cs]: OR 
```C#
DateTime.Now.Date.Subtract(TimeSpan.FromDays(7))
```

---
In reply to: [102630093](https://github.com/kuhlenh/Productivity/pull/1#discussion_r102630093) [](ancestors = 102630093)

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 35]
```diff
@@ -0,0 +1,258 @@
...
+        {
+            var lastWeekWorkouts = Workouts.Where(w => w.Date > DateTime.Now.Date.AddDays(-7));
+            var workoutWithMostCalsBurned = lastWeekWorkouts.Aggregate((w1, w2) => GetCaloriesBurned(w1) > GetCaloriesBurned(w2) ? w1 : w2);
+            return (workoutWithMostCalsBurned, GetCaloriesBurned(workoutWithMostCalsBurned));
+        }
+
+        public string TweetBestWorkoutOfWeek()
+        {
+            var best = GetWeeksBestWorkout();
+            if (best.Item2 > 200)
```
Comment (jmarolf) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: >Item2 [](start = 21, length = 5)

use tuple item name `best.calories`

File: src/BikeSharing.DomainLogic/BikeWorkout.cs [hunk 36]
```diff
@@ -0,0 +1,258 @@
...
+            {
+                return Tweetify(best.Item1.Notes);
```
Comment (jmarolf) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: >Item1 [](start = 37, length = 5)

use tuple item name `best.workout`

File: src/BikeSharing.DomainLogic/BikeWorkout.cs (hunk 35 above)
Comment (jmarolf) [inline src/BikeSharing.DomainLogic/BikeWorkout.cs]: >200 [](start = 29, length = 3)

As Cyrus says, have these be constants

File: src/BikeSharing.DomainLogic/BikeWorkout.cs (hunk 36 above)
Comment (kuhlenh) [reply src/BikeSharing.DomainLogic/BikeWorkout.cs]: more demo magic

File: src/BikeSharing.DomainLogic/BikeWorkout.cs (hunk 33 above)
Comment (kuhlenh) [reply src/BikeSharing.DomainLogic/BikeWorkout.cs]: treadmills! bicycles! that's how paces are calculated on those...

File: src/BikeSharing.DomainLogic/BikeWorkout.cs (hunk 26 above)
Comment (kuhlenh) [reply src/BikeSharing.DomainLogic/BikeWorkout.cs]: well i have to have a default...

**Pull request merged**